### PR TITLE
Milestone: Finished monkey-patching Enumerable module with custom methods

### PR DESCRIPTION
**Pull request by cyonii**
We created a our own custom enumerable methods and monkey-patched them to Ruby Enumerable module

Below are the methods created
- `#my_each`
- `#my_each_with_index`
- `#my_select`
- `#my_all?`
- `#my_any?`
- `#my_none?`
- `#my_count`
- `#my_map`
- `#my_inject`

### CODE REVIEW

**Review by mahmoud717 — CHANGES_REQUESTED** [review]
hey @cyonii & @dipbazz , Great work so far but there are still some changes to be made.
Please correct all the linters errors and file another review request.
Hint: use the command [rubocop -a] to fix these minor issues.
if you need any help please contact me on slack @mahmoud abbas

File: main.rb [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    return true if !to_check and length
+
+    my_each_with_index do |item, index|
+      next unless self[index + 1]
+
+      return false unless item == self[index + 1]
+    end
+    true
+  end
+
```
Comment (umairarshadbutt) [inline main.rb]: ### `my_all?` should be identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F).

File: main.rb [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+  def my_any?(to_check = nil)
+    return true if !to_check and length
+
+    my_each do |item|
+      return true if item == to_check
+    end
+    false
+  end
+
```
Comment (umairarshadbutt) [inline main.rb]: ### `my_any?` should be identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F).

File: main.rb [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+  def my_none?
+    return to_enum(:my_none?) unless block_given?
+
+    my_each do |item|
+      return false if yield(item)
+    end
+    true
+  end
+
```
Comment (umairarshadbutt) [inline main.rb]: ### `my_none?` should be identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F).

File: main.rb [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+  def my_count
+    counter = 0
+
+    my_each do
+      counter += 1
+    end
+
+    counter
+  end
+
```
Comment (umairarshadbutt) [inline main.rb]: ### `my_count?` should be identical and returns the same thing as ruby's [`count?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count).

File: main.rb [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+  def my_map(&my_proc)
+    return to_enum(:my_map) unless block_given?
+
+    map_list = []
+
+    my_each { |item| map_list.push(my_proc.call(item)) }
+    map_list
+  end
+
```
Comment (umairarshadbutt) [inline main.rb]: ### `my_map?` should be identical and returns the same thing as ruby's [`map?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map).

File: main.rb [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    counter = initial ? 0 : 1
+
+    my_each do
+      total = yield(total, self[counter]) if self[counter]
+      counter += 1
+    end
+    total
+  end
+end
+
```
Comment (umairarshadbutt) [inline main.rb]: ### `my_inject?` should be identical and returns the same thing as ruby's [`inject?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject).

File: main.rb [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+puts "\nRunning my_map method:"
+new_proc = proc { |item| item.to_s + 2.to_s }
+p my_array.my_map(&new_proc)
+
+puts "\n---------------------------------\n"
+puts "\nRunning my_inject method:"
+p [2, 4, 5].my_inject(nil) { |sum, n| sum * n }
+
+puts "\n---------------------------------\n"
+puts "\nRunning multiply_els method to test the output from my_inject method:"
```
Comment (umairarshadbutt) [inline main.rb]: Remove the `puts`

**Review by umairarshadbutt — CHANGES_REQUESTED** [review]
# Status: REQUIRED CHANGES :warning: :construction: 

Hi @dipbazz and @cyonii
You have done a splendid job with the changes! :tada: Keep it up.
We, however, do have a few more changes to carry out before we proceed (see review comment).

##
- [x] Please rename `main.rb` to `enumerables.rb` and remove the executable flag from the file.
- [x] Don't use `puts`
- [x] Add the MIT License in the README file. (MIT License is missing in the repository)
- [x] You have 29 failures please fix the failures

##
- **When you have made the specified changes, please submit another code review. Thank you!**
##
**If you are stuck at any point or have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help**
HAPPY CODING :computer:
**[Umair Arshad](https://umairarshad.me/)**
[Twitter](https://twitter.com/its_UmairArshad) | [LinkedIn](https://www.linkedin.com/in/umair-arshad-butt/) | [GitHub](https://github.com/umairarshadbutt)

File: main.rb (hunk 3 above)
Comment (cyonii) [reply main.rb]: The issue has been resolved

File: main.rb (hunk 4 above)
Comment (cyonii) [reply main.rb]: Implemented

File: enumerable.rb [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+  end
+
+  def my_map(&my_proc)
+    return to_enum(:my_map) unless block_given?
+
+    map_list = []
+
+    my_each { |item| map_list.push(my_proc.call(item)) }
+    map_list
+  end
```
Comment (maelfosso) [inline enumerable.rb]:
```rb
2.6.5 :009 > array
 => [4, 5, 4, 0, 8, 4, 5, 2, 4, 5, 4, 2, 3, 4, 3, 6, 5, 0, 0, 3, 6, 6, 4, 8, 1, 4, 5, 3, 2, 2, 8, 0, 3, 8, 0, 3, 4, 4, 8, 1, 0, 7, 7, 1, 2, 8, 6, 0, 8, 7, 1, 2, 4, 6, 2, 8, 6, 2, 1, 4, 4, 0, 6, 2, 6, 4, 4, 7, 5, 2, 2, 5, 7, 1, 2, 5, 1, 5, 7, 5, 1, 3, 3, 0, 7, 2, 1, 4, 7, 2, 5, 6, 0, 2, 6, 5, 2, 1, 2, 6] 
2.6.5 :010 > my_proc = proc { |num| num > 10 }
2.6.5 :011 > array.my_map(my_proc) { |num| num < 10 }
Traceback (most recent call last):
        7: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `<main>'
        6: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `eval'
        5: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `<main>'
        4: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `load'
        3: from /home/maelfosso/.rvm/gems/ruby-2.6.5/gems/irb-1.2.4/exe/irb:11:in `<top (required)>'
        2: from (irb):11
        1: from enumerable.rb:104:in `my_map'
ArgumentError (wrong number of arguments (given 1, expected 0))
2.6.5 :012 > array.map(&my_proc)
 => [false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false] 
2.6.5 :013 > 
```
it must execute only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.

File: enumerable.rb [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+
+  def my_inject(initial = nil)
+    total = initial
+
+    my_each do |value|
+      total = initial ? yield(total, value) : initial = value
+    end
+    total
+  end
+end
```
Comment (maelfosso) [inline enumerable.rb]:
```rb
2.6.5 :015 > array
 => [4, 5, 4, 0, 8, 4, 5, 2, 4, 5, 4, 2, 3, 4, 3, 6, 5, 0, 0, 3, 6, 6, 4, 8, 1, 4, 5, 3, 2, 2, 8, 0, 3, 8, 0, 3, 4, 4, 8, 1, 0, 7, 7, 1, 2, 8, 6, 0, 8, 7, 1, 2, 4, 6, 2, 8, 6, 2, 1, 4, 4, 0, 6, 2, 6, 4, 4, 7, 5, 2, 2, 5, 7, 1, 2, 5, 1, 5, 7, 5, 1, 3, 3, 0, 7, 2, 1, 4, 7, 2, 5, 6, 0, 2, 6, 5, 2, 1, 2, 6] 
2.6.5 :016 > array.inject(:+)
 => 378 
2.6.5 :017 > array.my_inject(:+)
Traceback (most recent call last):
       11: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `<main>'
       10: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `eval'
        9: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `<main>'
        8: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `load'
        7: from /home/maelfosso/.rvm/gems/ruby-2.6.5/gems/irb-1.2.4/exe/irb:11:in `<top (required)>'
        6: from (irb):17
        5: from enumerable.rb:116:in `my_inject'
        4: from enumerable.rb:7:in `my_each'
        3: from enumerable.rb:7:in `each'
        2: from enumerable.rb:8:in `block in my_each'
        1: from enumerable.rb:117:in `block in my_inject'
LocalJumpError (no block given (yield))
2.6.5 :020 > range = Range.new(5, 50)
2.6.5 :021 > range.inject(:+)
 => 1265 
2.6.5 :022 > range.my_inject(:+)
Traceback (most recent call last):
       11: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `<main>'
       10: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `eval'
        9: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `<main>'
        8: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `load'
        7: from /home/maelfosso/.rvm/gems/ruby-2.6.5/gems/irb-1.2.4/exe/irb:11:in `<top (required)>'
        6: from (irb):22
        5: from enumerable.rb:116:in `my_inject'
        4: from enumerable.rb:7:in `my_each'
        3: from enumerable.rb:7:in `each'
        2: from enumerable.rb:8:in `block in my_each'
        1: from enumerable.rb:117:in `block in my_inject'
LocalJumpError (no block given (yield))
2.6.5 :023 > 
```

Comment (maelfosso) [reply enumerable.rb]:
```rb
2.6.5 :025 > array
 => [4, 5, 4, 0, 8, 4, 5, 2, 4, 5, 4, 2, 3, 4, 3, 6, 5, 0, 0, 3, 6, 6, 4, 8, 1, 4, 5, 3, 2, 2, 8, 0, 3, 8, 0, 3, 4, 4, 8, 1, 0, 7, 7, 1, 2, 8, 6, 0, 8, 7, 1, 2, 4, 6, 2, 8, 6, 2, 1, 4, 4, 0, 6, 2, 6, 4, 4, 7, 5, 2, 2, 5, 7, 1, 2, 5, 1, 5, 7, 5, 1, 3, 3, 0, 7, 2, 1, 4, 7, 2, 5, 6, 0, 2, 6, 5, 2, 1, 2, 6] 
2.6.5 :026 > array.inject(20, :+)
 => 398 
2.6.5 :027 > array.my_inject(20, :+)
Traceback (most recent call last):
        7: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `<main>'
        6: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `eval'
        5: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `<main>'
        4: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `load'
        3: from /home/maelfosso/.rvm/gems/ruby-2.6.5/gems/irb-1.2.4/exe/irb:11:in `<top (required)>'
        2: from (irb):27
        1: from enumerable.rb:113:in `my_inject'
ArgumentError (wrong number of arguments (given 2, expected 0..1))
2.6.5 :028 > range.inject(20, :+)
 => 1285 
2.6.5 :029 > range.my_inject(20, :+)
Traceback (most recent call last):
        7: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `<main>'
        6: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `eval'
        5: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `<main>'
        4: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `load'
        3: from /home/maelfosso/.rvm/gems/ruby-2.6.5/gems/irb-1.2.4/exe/irb:11:in `<top (required)>'
        2: from (irb):29
        1: from enumerable.rb:113:in `my_inject'
ArgumentError (wrong number of arguments (given 2, expected 0..1))
2.6.5 :030 > 
```

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required 🔴 

Hello,

Thank you for implementing the suggested changes.
However, there are still some missing requirements.

1- Kindly read the comments below to know more about the bugs to fix

2- The `rubocop.yml` file in your project is not identical to the required one. Kindly follow that [llink](https://github.com/microverseinc/linters-config/ruby) to set up the good one

Once all the changes are made, kindly submit a new code review request
Thanks

Mael FOSSO

**Comment by cyonii** [comment]
Thank you for providing test cases

**Review by PhillipUg — CHANGES_REQUESTED** [review]
# Changes Required ♻️ 

Impressive 🥳 almost all your methods pass the provided test cases 

Just one remaining

- [x] Your `my_inject` method still needs some work. Please make sure it passes the test cases provided below.

```
p [1, 2, 3, 4].my_inject(10) { |accum, elem| accum + elem } # => 20
p [1, 2, 3, 4].my_inject { |accum, elem| accum + elem } # => 10
p [5, 1, 2].my_inject('+') # => 8
p (5..10).my_inject(2, :*) # should return 302400
p (5..10).my_inject(4) { |prod, n| prod * n } # should return 604800
```

Please make the suggested changes and resubmit for review. Otherwise keep up the good work. ✔️ 

Happy coding 😎 

For any assistance feel free to reach out to me on slack as **Phillip Musiime**

**Review by Georjane — APPROVED** [review]
# Status: PR Approved ✔️
## Hi @cyonii and @dipbazz 
Nice work with the enumerable methods. 👏 👏 👏 Your my_inject now works as required.
This PR has been approved and you can go ahead and merge it. 👍

I wish you the very best in your next project, and Happy Coding 💻

**Pull request merged**